design/baremetal-operator/configmap_for_host_inspection_data.md

```diff
@@ -0,0 +1,193 @@
+- Adjust pivoting workflow to restore inspection data from HostData
+- Make the HostData CR immutable
+- Don't write `status.hardware` data to the status annotation
+
+### Non-Goals
+
+- Remove `status.hardware` from BareMetalHost CR
+- Drop status annotation
+
+## Proposal
+
+### User Stories
+
+As an operator, from an ephemeral cluster I would like to pivot/move all the Cluster
+API, CAPM3 and BMO objects to a target cluster successfully.
+
+## Design Details
+
+Apart from writing hardware details into the status of BareMetalHost, write it to a
+HostData CR spec. BareMetalHost and HostData will be parent and child objects
+respectively, and this relationship is controlled via ownerReferences chain. As such,
+the same name & namespace will be used for HostData as its owning object BareMetalHost.
+The spec of HostData will contain the same fields & hardware information as currently
+under `status.hardware` of BareMetalHost.
+
+### Pivoting
+
+To pivot objects excluded from Cluster API chain to a target cluster, one can set
+`clusterctl.cluster.x-k8s.io=""` label on HostData and BareMetalHost CRDs.
+Current status annotation based pivoting workflow will be stay as it is. However,
+CAPM3 machine controller will no longer copy `status.hardware` of BareMetalHost
+to the annotation. Because that part will be avialable in the form of HostData CR
+in the target cluster.
+
+### Implementation Details/Notes/Constraints
+
+The life of HostData is dependent on BareMetalHost. When BareMetalHost is entering
+inspection state, the HostData will be created. Failure in reaching inspection state
+or having `inspect.metal3.io: disabled` annotation on BareMetalHost will force the
+operator to not create the HostData. Deleting a BareMetalHost will result in deletion
+of a HostData respectively.
+
+HostData should be immutable and this will be controlled via [Validating Webhook](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#validatingadmissionwebhook) and with RBAC permissions.
+In case, when [re-inspection](https://github.com/metal3-io/baremetal-operator/blob/master/docs/inspectAnnotation.md) of a `available` BareMetalHost is requested,
+new HostData will be re-created, because updating the old HostData will be
+blocked by the Validating Webhook.
+
+Deprovisioning of the BareMetalHost must not delete the HostData. It can
+be deleted either when BareMetalHost is being deleted or when re-inspection
+is requested. To ensure HostData availability while BareMetalHost is provisioned,
+a finalizer should be set on the HostData to prevent its accidental deletion.
+
+Only operator will be granted with create permissions for HostData and everyone
+else will be given no write permissions. If the BareMetalHost was inspected but
+HostData object doesn't exist, operator will first try to copy it from the
+BareMetalHost status (if available) and vice versa. If not available in neither
+of them, new inspection will be enforced.
```

Comment:
A finalizer can always be manually patched to force a deletion at any time. This could generate a situation where an already provisioned (or any other state after inspection) host could suddenly loose the inspection data. Forcing re-inspection could become tricky (and this could be more problematic when the field will be removed from the host status)

Reply:
That's right. However, I think in general we give cluster access to someone we trust. For example, accidental deletion of BMH can also happen if the cluster access is not secured. So, perhaps we will need to clearly mention that in the user document as potential cause of not securing the cluster?

Also, that's how Cluster API workflow is also designed. For example, if someone intentionally/accidentally deletes Cluster object, or Machine, then it will result in other objects deletion/deprovisioning on the chain.

Reply:
More than a security problem I think that probably the `new inspection will be enforced` phrase could be detailed a little bit more. In general, if the HostData deletion will happen after the `Inspecting` and within the `Available` state, the host could be forced back to the `Inspecting` without any particular issue. Instead, if it happens during a `Provisioned` state, as you also already mentioned, this could  trigger a deprovisioning first